### PR TITLE
Refactored setupapi parser #2717

### PR DESCRIPTION
**Pull request by Onager**
Also changed the EncodedTextReader to not strip every line and updated parsers that relied on this behavior.

There's also more changes that I'd like make to the parser, but I've tried to keep changes to the tests minimal in this first round.

### CODE REVIEW

File: plaso/parsers/setupapi.py [hunk 1]
```diff
@@ -63,74 +65,109 @@ class SetupapiLogParser(text_parser.PyparsingMultiLineTextParser):
...
+  def _BuildDatetime(self, time_structure):
+    """Builds time elements from a setupapi time stamp.
 
     Args:
-      parser_mediator (ParserMediator): mediates interactions between parsers
-          and other components, such as storage and dfvfs.
-      structure (pyparsing.ParseResults): structure of tokens derived from
-          log entry.
+      time_structure (pyparsing.ParseResults): structure of tokens derived from
+          a setupapi time stamp.
```
Comment (joachimmetz) [inline plaso/parsers/setupapi.py]: time stamp => date and time value

File: plaso/parsers/setupapi.py [hunk 2]
```diff
@@ -63,74 +65,109 @@ class SetupapiLogParser(text_parser.PyparsingMultiLineTextParser):
...
+
+    Returns:
+      dfdatetime.TimeElements: date and time extracted from the value or None
```
Comment (joachimmetz) [inline plaso/parsers/setupapi.py]: the value => the structure

File: plaso/parsers/setupapi.py [hunk 3]
```diff
@@ -63,74 +65,109 @@ class SetupapiLogParser(text_parser.PyparsingMultiLineTextParser):
...
+          if the value does not represent a valid string.
```
Comment (joachimmetz) [inline plaso/parsers/setupapi.py]: a valid string => a valid date and time value

File: plaso/parsers/setupapi.py [hunk 4]
```diff
@@ -145,47 +182,78 @@ def ParseRecord(self, parser_mediator, key, structure):
...
           and other components, such as storage and dfvfs.
-      lines (str): one or more lines from the text file.
+      line (str): single line from the text file.
 
     Returns:
       bool: True if this is the correct parser, False otherwise.
     """
     try:
-      structure = self._SETUPAPI_LINE.parseString(lines)
+      _ = self._LOG_HEADER_START.parseString(line)
```
Comment (joachimmetz) [inline plaso/parsers/setupapi.py]: remove `_ =` not needed

File: plaso/parsers/winfirewall.py [hunk 5]
```diff
@@ -224,7 +224,7 @@ def VerifyStructure(self, parser_mediator, line):
     """
     # TODO: Examine other versions of the file format and if this parser should
     # support them.
-    return line == '#Version: 1.5'
+    return '#Version: 1.5' in line
```
Comment (joachimmetz) [inline plaso/parsers/winfirewall.py]: Why change this?

File: tests/parsers/setupapi.py [hunk 6]
```diff
@@ -101,8 +101,8 @@ def testParseSetupLog(self):
 
     self.CheckTimestamp(event.timestamp, '2015-11-22 17:53:29.305000')
 
-    expected_message = ('Setup Plug and Play Device Install - START')
-    expected_short_message = ('START - Setup Plug and Play Device Install')
+    expected_message = ('Setup Plug and Play Device Install')
```
Comment (joachimmetz) [inline tests/parsers/setupapi.py]: remove bounding parenthesis

File: plaso/parsers/setupapi.py (hunk 1 above)
Comment (Onager) [reply plaso/parsers/setupapi.py]: The setupapi documents call this field a timestamp:  https://docs.microsoft.com/en-us/windows-hardware/drivers/install/format-of-a-text-log-section-header

Clarified this.

File: plaso/parsers/setupapi.py (hunk 2 above)
Comment (Onager) [reply plaso/parsers/setupapi.py]: Done.

File: plaso/parsers/setupapi.py (hunk 3 above)
Comment (Onager) [reply plaso/parsers/setupapi.py]: Done.

File: plaso/parsers/setupapi.py (hunk 4 above)
Comment (Onager) [reply plaso/parsers/setupapi.py]: Done.

File: tests/parsers/setupapi.py (hunk 6 above)
Comment (Onager) [reply tests/parsers/setupapi.py]: Done.

File: plaso/parsers/winfirewall.py (hunk 5 above)
Comment (Onager) [reply plaso/parsers/winfirewall.py]: Because of the whitespace matching. Changed this to just have the newline.

**Comment by Onager** [comment]
@js-forensic - good idea, I thought this had already been done

File: plaso/parsers/setupapi.py [hunk 7]
```diff
@@ -145,47 +182,78 @@ def ParseRecord(self, parser_mediator, key, structure):
...
+      if not end_time:
+        parser_mediator.ProduceExtractionWarning(
+            'invalid date time value: {0!s}'.format(time_structure))
+      self._last_end_time = end_time
+      return
+
+    if key == 'section_end_exit_status':
+      exit_status = self._GetValueFromStructure(
+          structure, 'exit_status')
+      if self._last_end_time:
```
Comment (joachimmetz) [inline plaso/parsers/setupapi.py]: Please add a comment that explains why exit time is stored internally first and than later used

File: plaso/parsers/setupapi.py [hunk 8]
```diff
@@ -145,47 +182,78 @@ def ParseRecord(self, parser_mediator, key, structure):
     Raises:
       ParseError: when the structure type is unknown.
     """
-    if key != 'logline':
-      raise errors.ParseError(
-          'Unable to parse record, unknown structure: {0:s}'.format(key))
+    if key == 'ignorable_line':
```
Comment (joachimmetz) [inline plaso/parsers/setupapi.py]: what if self._last_end_time contains a value from the previous file it parsed?

File: plaso/parsers/winfirewall.py [hunk 9]
```diff
@@ -224,7 +224,8 @@ def VerifyStructure(self, parser_mediator, line):
     """
     # TODO: Examine other versions of the file format and if this parser should
     # support them.
-    return line == '#Version: 1.5'
+    stripped_line = line.strip()
```
Comment (joachimmetz) [inline plaso/parsers/winfirewall.py]: strip() => rstrip()

File: plaso/parsers/setupapi.py (hunk 8 above)
Comment (Onager) [reply plaso/parsers/setupapi.py]: Added code to reset these values to VerifyStructure.

File: plaso/parsers/setupapi.py (hunk 7 above)
Comment (Onager) [reply plaso/parsers/setupapi.py]: Done

File: plaso/parsers/winfirewall.py (hunk 9 above)
Comment (Onager) [reply plaso/parsers/winfirewall.py]: Done

**Comment by joachimmetz** [comment]
tox tests are failing:
```
testExamineEventAndCompileReportOnSystemFile (analysis.windows_services.WindowsServicesTest)
1614Tests the ExamineEvent and CompileReport functions on a SYSTEM file. ... free(): invalid next size (fast)
1615ERROR: InvocationError: '/plaso/.tox/py27/bin/coverage run --source=plaso --omit=*_test*,*__init__*,*test_lib* run_tests.py'
```

**Comment by Onager** [comment]
Ping for @joachimmetz

File: plaso/parsers/mac_wifi.py [hunk 10]
```diff
@@ -229,14 +229,16 @@ def _ParseLogLine(self, parser_mediator, key, structure):
 
     function = self._GetValueFromStructure(structure, 'function')
     if function:
-      # Due to the use of CharsNotIn the function value can contain leading
```
Comment (joachimmetz) [inline plaso/parsers/mac_wifi.py]: why remove this?

File: plaso/parsers/setupapi.py [hunk 11]
```diff
@@ -63,74 +65,109 @@ class SetupapiLogParser(text_parser.PyparsingMultiLineTextParser):
...
+
+  def __init__(self):
+    """Initializes a setupapi parser."""
+    super(SetupapiLogParser, self).__init__()
+    self._last_end_time = None
+    self._last_entry_type = None
 
-  def _ParseRecordLogline(self, parser_mediator, structure):
-    """Parses a logline record structure and produces events.
+  def _BuildDatetime(self, time_structure):
```
Comment (joachimmetz) [inline plaso/parsers/setupapi.py]: Change name of this method to `_GetDateTime` or equiv to keep it consistent with other parsers

File: plaso/parsers/mac_wifi.py (hunk 10 above)
Comment (Onager) [reply plaso/parsers/mac_wifi.py]: CharsNotIn is not used in the function definition. Removed the strip here altogether, as it's not necessary.

File: plaso/parsers/setupapi.py (hunk 11 above)
Comment (Onager) [reply plaso/parsers/setupapi.py]: Done

File: plaso/parsers/mac_wifi.py (hunk 10 above)
Comment (joachimmetz) [reply plaso/parsers/mac_wifi.py]: It is used in the pyparsing grammer, https://github.com/log2timeline/plaso/blob/e5808cb4b35104b00db2f344d3c6a036e328b000/plaso/parsers/mac_wifi.py#L69

Comment (joachimmetz) [reply plaso/parsers/mac_wifi.py]: per conversation, this applies to agent not function and text.

**Pull request merged**